#Converting user comments data into SQL table

In [29]:
import sys
sys.path.append('/Users/alexsutherland/Dropbox/Insight/WeightLossReddit/')
import redditDataIO
import pandas as pd
import numpy as np
import os
os.chdir('/Users/alexsutherland/Dropbox/Insight/WeightLossReddit/')

Loading the data from a csv:

In [5]:
userCommentData = pd.read_csv('Data/6.20.15-UserComment.csv')

/Users/alexsutherland/anaconda/lib/python2.7/site-packages/pandas/io/parsers.py:1139: DtypeWarning: Columns (0,1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  data = self._reader.read(nrows)


Let's look to see if there's anything we can clean up:

In [6]:
userCommentData.head()

Unnamed: 0 score controversiality  \
0          0     2                0   
1          1    11                0   
2          2     2                0   
3          3     1                0   
4          4     3                0   

                                                body   subreddit  \
0  Lenny is the man. I'm good friends with his so...       funny   
1                           CondoInternet, brobeans.  technology   
2                   Baller, thanks!! Happy new year!     Seattle   
3                    Awesome, thanks for the advice!     Seattle   
4  Great call! I'm from a place with 0 public tra...     Seattle   

                                          link_title    link_id  created_utc  \
0                   Millionaires In America (Not OC)  t3_2u8mg2   1422674804   
1  SpaceX confirms $1 billion funding round with ...  t3_2t3giy   1421824107   
2  First New Year's Eve in Seattle, and I have a ...  t3_2qwaqb   1420078827   
3  First New Year's Eve in Seattle, and I have a ...  t3_2qwaqb   1420048144   
4  First New Year's Eve in Seattle, and I have a ...  t3_2qwaqb   1420008479   

        username  userid  
0  BrovietUnion1       1  
1  BrovietUnion1       1  
2  BrovietUnion1       1  
3  BrovietUnion1       1  
4  BrovietUnion1       1

The first column is not useful, let's drop that. 

In [11]:
userCommentDataCleaned = userCommentData.drop(['Unnamed: 0'],axis=1)
userCommentDataCleaned.head()

score controversiality                                               body  \
0     2                0  Lenny is the man. I'm good friends with his so...   
1    11                0                           CondoInternet, brobeans.   
2     2                0                   Baller, thanks!! Happy new year!   
3     1                0                    Awesome, thanks for the advice!   
4     3                0  Great call! I'm from a place with 0 public tra...   

    subreddit                                         link_title    link_id  \
0       funny                   Millionaires In America (Not OC)  t3_2u8mg2   
1  technology  SpaceX confirms $1 billion funding round with ...  t3_2t3giy   
2     Seattle  First New Year's Eve in Seattle, and I have a ...  t3_2qwaqb   
3     Seattle  First New Year's Eve in Seattle, and I have a ...  t3_2qwaqb   
4     Seattle  First New Year's Eve in Seattle, and I have a ...  t3_2qwaqb   

   created_utc       username  userid  
0   1422674804  BrovietUnion1       1  
1   1421824107  BrovietUnion1       1  
2   1420078827  BrovietUnion1       1  
3   1420048144  BrovietUnion1       1  
4   1420008479  BrovietUnion1       1

In [15]:
userCommentDataCleaned.describe()

created_utc         userid
count  9.734580e+05  973458.000000
mean   1.394449e+09    1557.345064
std    3.160666e+07     958.313796
min    1.156315e+09       1.000000
25%    1.375850e+09     748.000000
50%    1.400912e+09    1514.000000
75%    1.420354e+09    2350.000000
max    1.434830e+09    3341.000000

For some reason, the score and controversiality columns are now showing up in the .describe() table. Let's investigate why:

In [14]:
userCommentDataCleaned['score'].value_counts()

1     351474
2     172968
3      80593
1      53513
4      29696
0      28278
5      27079
2      26779
6      19845
7      14491
3      12879
8      11082
9       8802
-1      8235
10      7142
...
2175     1
1050     1
1053     1
1055     1
1057     1
2370     1
664      1
765      1
trees    1
3544     1
768      1
761      1
767      1
764      1
1054     1
Length: 2057, dtype: int64

It looks like there is some text in this column:

In [16]:
userCommentDataCleaned.ix[userCommentDataCleaned['score'] == 'trees',:]

score                    controversiality       body     subreddit  \
374706  trees  My smoke spot for tonight. Go Avs!  t3_24f1fg  1398921541.0   

         link_title link_id  created_utc username  userid  
374706  JimmyTree46    1160          NaN      NaN     NaN

Let's drop this row to see if that improves things:

In [17]:
userCommentDataCleaned = userCommentDataCleaned.drop(374706)

In [51]:
userCommentDataCleaned2 = userCommentDataCleaned.copy()
for item in userCommentDataCleaned['score'].value_counts().index:
    try:
        int(item)
    except:
        userCommentDataCleaned2 = userCommentDataCleaned2.drop(userCommentDataCleaned2.ix[userCommentDataCleaned2['score'] == item].index.values)

In [52]:
userCommentDataCleaned.shape

(973795, 9)

In [53]:
userCommentDataCleaned2.shape

(973690, 9)

In [55]:
userCommentDataCleaned2.dtypes

score                object
controversiality     object
body                 object
subreddit            object
link_title           object
link_id              object
created_utc         float64
username             object
userid              float64
dtype: object

It still thinks score is an object, not an int/float. Let's try to convert this:

In [58]:
userCommentDataCleaned2['score'] = userCommentDataCleaned2['score'].convert_objects(convert_numeric=True)

In [59]:
userCommentDataCleaned2.dtypes

score               float64
controversiality     object
body                 object
subreddit            object
link_title           object
link_id              object
created_utc         float64
username             object
userid              float64
dtype: object

In [60]:
userCommentDataCleaned2.describe()

score   created_utc         userid
count  973564.000000  9.734580e+05  973458.000000
mean        6.809339  1.394449e+09    1557.345064
std        53.149711  3.160666e+07     958.313796
min      -245.000000  1.156315e+09       1.000000
25%         1.000000  1.375850e+09     748.000000
50%         2.000000  1.400912e+09    1514.000000
75%         3.000000  1.420354e+09    2350.000000
max      6014.000000  1.434830e+09    3341.000000

Great, score is now a 64 bit float. Let's look into why controversiality is not a float:

In [61]:
userCommentDataCleaned2['controversiality'].value_counts()

0    827897
0    128658
1     14595
1      2414
dtype: int64

Looks like an issue with some being ints and others being strings, let's do another convert objects:

In [62]:
userCommentDataCleaned2['controversiality'] = userCommentDataCleaned2['controversiality'].convert_objects(convert_numeric=True)

In [63]:
userCommentDataCleaned2.describe()

score  controversiality   created_utc         userid
count  973564.000000     973564.000000  9.734580e+05  973458.000000
mean        6.809339          0.017471  1.394449e+09    1557.345064
std        53.149711          0.131018  3.160666e+07     958.313796
min      -245.000000          0.000000  1.156315e+09       1.000000
25%         1.000000          0.000000  1.375850e+09     748.000000
50%         2.000000          0.000000  1.400912e+09    1514.000000
75%         3.000000          0.000000  1.420354e+09    2350.000000
max      6014.000000          1.000000  1.434830e+09    3341.000000

Looks good. Let's look at the full table again:

In [65]:
userCommentDataCleaned2

score  controversiality  \
0           2                 0   
1          11                 0   
2           2                 0   
3           1                 0   
4           3                 0   
5           1                 0   
6           2                 0   
7           2                 0   
8           2                 0   
9           1                 0   
10          2                 0   
11          2                 0   
12          2                 0   
13          1                 0   
14          2                 0   
15          3                 0   
16          1                 0   
17          3                 0   
18          2                 0   
19          3                 0   
20          3                 0   
21          8                 0   
22         15                 0   
23          3                 0   
24          5                 0   
25          8                 0   
26          3                 0   
27          9                 0   
28         28                 0   
29          8                 0   
...       ...               ...   
973766      0                 0   
973767      0                 0   
973768      0                 0   
973769      0                 0   
973770      1                 0   
973771      0                 0   
973772      0                 0   
973773      0                 0   
973774      2                 1   
973775     -1                 0   
973776      1                 0   
973777      1                 0   
973778      1                 0   
973779      1                 0   
973780     -5                 0   
973781      1                 0   
973782      2                 0   
973783      1                 0   
973784      1                 0   
973785      1                 0   
973786      1                 0   
973787      1                 0   
973788      1                 0   
973789      1                 0   
973790      1                 0   
973791      1                 0   
973792      1                 0   
973793      1                 0   
973794      3                 0   
973795      4                 0   

                                                     body  \
0       Lenny is the man. I'm good friends with his so...   
1                                CondoInternet, brobeans.   
2                        Baller, thanks!! Happy new year!   
3                         Awesome, thanks for the advice!   
4       Great call! I'm from a place with 0 public tra...   
5       That number was from when I was cutting 2-3 lb...   
6       Haha well thanks for the ideas. This thread wa...   
7       Sounds super fun! I'd definitely be down haha ...   
8                                      I credit hot broga   
9       I like the "no table separating you" bit. I ne...   
10                   Haha you sound like an awesome date!   
11      That sounds awesome, actually! Guessing it's n...   
12      Sounds fun! What if they suck at ice skating /...   
13      Sure thing, and thanks!\r\n\r\nThe studio I go...   
14      Totally true, the diet is so key. I'm making t...   
15      Haven't been on nearly as good of behavior as ...   
16      Haha there wasn't a huge August - September di...   
17      The pics represent August, October, November, ...   
18      [Find something cool to do at home!](http://ww...   
19      Actually, no. How do I measure that? Happy to ...   
20      Hmmm. Strange that your routine isn't easy but...   
21      Thanks! This [first thread](http://www.reddit....   
22      Haha my girlfriend and I have been dating for ...   
23      Thanks!! And I know, it wasn't a terrible star...   
24      Lol I *don't* make enough money... at the mome...   
25      /u/ohrejoyce is pretty much right. I read thro...   
26        I know :) not sure who downvoted you though :-/   
27      Haha thanks, she's not so bad herself though.\...   
28      Thanks!\r\n\r\nI put on weight over a few mont...   
29     

In [66]:
userCommentDataCleaned2['subreddit'].value_counts()

AskReddit          107219
progresspics        35190
loseit              29264
funny               25373
pics                24968
AdviceAnimals       20350
Fitness             18635
keto                15031
WTF                 14239
MakeupAddiction     12146
videos              10504
gaming              10200
trees                8090
IAmA                 7942
nfl                  7558
...
mildlyenteristing        1
ecchi                    1
nerdblock                1
RandomActsOfPetFood      1
cfbball                  1
tmbg                     1
cheerleaders             1
OregonCoast              1
TechnicalDeathMetal      1
Sum41                    1
Techno                   1
thronescomics            1
colorador4r              1
aaaaaatheismmmmmmmmmm    1
StarWarsEU               1
Length: 9351, dtype: int64

Alright, let's output this into a sql table:

In [67]:
redditDataIO.exportDataToSQL(userCommentDataCleaned2, tableName='userCommentData')

Index([u'score', u'controversiality', u'body', u'subreddit', u'link_title', u'link_id', u'created_utc', u'username', u'userid'], dtype='object')


/Users/alexsutherland/anaconda/lib/python2.7/site-packages/pymysql/cursors.py:134: Warning: Unknown table 'redditweightdatabase.usercommentdata'
  result = self._query(query)


InternalError: (InternalError) (1054, u"Unknown column 'nan' in 'field list'") 'INSERT INTO `userCommentData` (`index`, score, controversiality, body, subreddit, link_title, link_id, created_utc, username, userid) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)' ((0, 2.0, 0.0, "Lenny is the man. I'm good friends with his son and met him on one occasion; he loves his role as batman and brings a lot of happiness to people under the radar. The guy is a legend in my circle of friends.", 'funny', 'Millionaires In America (Not OC)', 't3_2u8mg2', 1422674804.0, 'BrovietUnion1', 1.0), (1, 11.0, 0.0, 'CondoInternet, brobeans.', 'technology', 'SpaceX confirms $1 billion funding round with Google', 't3_2t3giy', 1421824107.0, 'BrovietUnion1', 1.0), (2, 2.0, 0.0, 'Baller, thanks!! Happy new year!', 'Seattle', "First New Year's Eve in Seattle, and I have a friend arriving at the airport at 8:30pm on NYE. Is it better for me to pick my friend up, or have them take a cab to me? Any tips/advice?", 't3_2qwaqb', 1420078827.0, 'BrovietUnion1', 1.0), (3, 1.0, 0.0, 'Awesome, thanks for the advice!', 'Seattle', "First New Year's Eve in Seattle, and I have a friend arriving at the airport at 8:30pm on NYE. Is it better for me to pick my friend up, or have them take a cab to me? Any tips/advice?", 't3_2qwaqb', 1420048144.0, 'BrovietUnion1', 1.0), (4, 3.0, 0.0, "Great call! I'm from a place with 0 public transport, so that didn't even cross my mind. We'll walk from downtown to cap hill, thanks so much!", 'Seattle', "First New Year's Eve in Seattle, and I have a friend arriving at the airport at 8:30pm on NYE. Is it better for me to pick my friend up, or have them take a cab to me? Any tips/advice?", 't3_2qwaqb', 1420008479.0, 'BrovietUnion1', 1.0), (5, 1.0, 0.0, "That number was from when I was cutting 2-3 lbs/week (if you see my previous posts from last year). I think a lot of people rightfully pointed out that I used to weight lift, so I had some muscle underneath the pudge. Last time, I really spent a ton of time just cutting weight, doing cardio, and low-weight-high-rep weightlifting to keep myself strong. This time, I think it's been more exercised-induced progress, and less about diet or calorie counting.\r\n\r\nThat said, I'm trying to get back into that &gt;500 calorie deficit/day range so I can drop 5 or so more pounds in the next month or two.", 'progresspics', 'Update [M/24/6\'0"]: 1.5 years later, got fat again! My girlfriend broke up with me 4 months ago, so I whipped myself back into shape: 195 -&gt; 185 lbs over 4 months of being single, plus muscle gain. The weight will come back, you have to keep at it!', 't3_2owbfv', 1418377352.0, 'BrovietUnion1', 1.0), (6, 2.0, 0.0, 'Haha well thanks for the ideas. This thread was helpful, going on date 2 with someone tomorrow night. Hopefully she likes video games!', 'AskWomen', 'I\'ve often wondered: What is your perfect "first, second, third" date combination? That is, what would you personally love to do and happen on your first 3 dates with someone new?', 't3_2p0cyo', 1418345620.0, 'BrovietUnion1', 1.0), (7, 2.0, 0.0, "Sounds super fun! I'd definitely be down haha sounds like date 3 may be about more than video games though ;)", 'AskWomen', 'I\'ve often wondered: What is your perfect "first, second, third" date combination? That is, what would you personally love to do and happen on your first 3 dates with someone new?', 't3_2p0cyo', 1418344135.0, 'BrovietUnion1', 1.0)  ... displaying 10 of 973690 total bound parameter sets ...  (973794, 3.0, 0.0, 'OG black Pebble, skin is GadgetWraps dark tamo, band is Richard Tracy wide body black matte leather (bought by my gf for my birthday).', 'pebble', '[Band][Skin] My Pebble customization, if anyone cares.', 't3_34bu4g', 1430345900.0, 'Iron_Raiden_Mecord', 3341.0), (973795, 4.0, 0.0, 'Do NOT let it go almost dead, whatever you do. That\'s a great way to dramatically shorten the lifespan of a Li-ion/LiPo battery, and since you can\'t really replace the battery on a Pebble, trashed battery = trashed watch. It\'s much better to just "top it off" whenever it\'s convenient. I usually do it around 50%, which takes a few days anyway.', 'pebble', 'At what battery percentage do you charge your pebble?', 't3_2r9yco', 1420355497.0, 'Iron_Raiden_Mecord', 3341.0))

Whoops, we need to remove any row that has NaNs

In [68]:
userCommentDataCleaned3 = userCommentDataCleaned2.dropna()

In [71]:
userCommentDataCleaned3.shape[0] - userCommentDataCleaned2.shape[0]

-251

That removed 251 rows that had NaN. With those gone, let's try exporting again:

In [73]:
redditDataIO.exportDataToSQL(userCommentDataCleaned3, tableName='userCommentData')

Index([u'score', u'controversiality', u'body', u'subreddit', u'link_title', u'link_id', u'created_utc', u'username', u'userid'], dtype='object')
Exported Data to SQL Table (Name: userCommentData)
